### PR TITLE
fix(committees): recover empty member project_uid for Org Lens seat reads (LFXV2-1865)

### PR DESCRIPTION
**Pull request by luismoriguerra**
## Summary

The Org Lens **Board & Committee** tab under-counts an organization's committee seats. For Red Hat × Agentic AI Foundation (AAIF) the card showed **2** committee seats while Snowflake and the query-service index both report **13 seats / 6 distinct people**.

**Root cause:** members created before `project_uid`/`project_slug` were denormalized onto committee_member records (LFXV2-1442) carry an **empty `project_uid`** in the committee-members KV truth. The org seat read (`committees/b2b-org/{uid}/seats`) filters an org's members by the member's own `project_uid` against the foundation project family, so those legacy seats are silently dropped. The OpenSearch index already recovers `project_uid`, which is why query-service shows all 13 but the KV-backed card shows only the 2 records written after the fix.

## Changes

- **Read fallback** (`internal/infrastructure/nats/org_committee_seat_reader.go`): when a member's `project_uid` is empty, recover it from the parent committee (cached per committee; a committee-read failure degrades to the prior exclusion rather than failing the whole list) and enrich the in-memory member. The tab now renders the full seat set with **no data migration required**. Wired via `providers.go` (`CommitteeBaseReader`).
- **Backfill** (`cmd/committee-cli/commands/sync/member-project-attribute`): reconciles member `project_uid`/`project_slug` against the parent committee so KV truth matches every consumer. Streams the bucket (`EachMember`), caches committee lookups, **defaults to `--dry-run=true`**, supports `--committee-uid` / `--project-uid` scoping and an interruptible `--sleep`, and is idempotent.
- **Regression test** (`internal/service/committee_member_writer_test.go`): asserts `CreateMember` denormalizes `project_uid`/`project_slug` from the committee **and overrides any caller-supplied value** — locking the invariant that previously shipped untested.

## Test plan

- [x] `gofmt -l`, `go vet ./...`, `golangci-lint run ./...` (v2.2.2) — clean
- [x] `go build ./...` — OK
- [x] `go test -count=1 ./...` — all packages pass
- [x] New seat-reader fallback tests (recovery, per-committee cache, lookup-failure-degrades-to-exclusion)
- [x] New `member-project-attribute` tests (repair+enrich, idempotent skip, `--project-uid` scoping, dry-run no-writes, committee-read-error, mutually-exclusive flags)
- [x] New `CreateMember` denormalization regression test
- [ ] Suggested rollout: run `committee-cli sync member-project-attribute --project-uid <aaif-project-uid>` (dry-run) then `--dry-run=false --sleep 200ms` to repair KV truth; drop `--project-uid` to heal all foundations provisioned before LFXV2-1442.

No Goa design or Heimdall ruleset changes. No new dependencies.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: cmd/committee-cli/commands/sync/members_by_committee_index_test.go [hunk 1]
```diff
@@ -20,15 +20,21 @@ import (
 // mockMemberWriter is a minimal implementation of port.CommitteeMemberWriter used by
 // the backfill subcommand.  Only IndexMemberByCommittee is exercised here.
 type mockMemberWriter struct {
-	indexed    []string // committee_uid+"."+member_uid keys that were written
-	orgIndexed []string // org_sfid+"."+member_uid keys that were written
-	indexError error
+	indexed     []string                 // committee_uid+"."+member_uid keys that were written
```
[the remark this answers is not in the document]
Comment (luismoriguerra) [reply cmd/committee-cli/commands/sync/members_by_committee_index_test.go]: Good catch — updated the doc comment on `mockMemberWriter` to note it's a full `port.CommitteeMemberWriter` shared across the sync backfill tests (members-by-committee, members-by-organization, member-project-attribute) and that `IndexMemberByCommittee`, `IndexMemberByOrganization`, and `UpdateMember` (call recording + error simulation) are all exercised. Comment text only, no behavior change. Fixed in fdd39a4.

**Comment by luismoriguerra** [comment]
### Review round recap

**Resolved (1):**
- Copilot — `cmd/committee-cli/commands/sync/members_by_committee_index_test.go:23`: stale doc comment on `mockMemberWriter`. Updated to accurately describe it as a full `port.CommitteeMemberWriter` shared across the sync backfill tests, exercising `IndexMemberByCommittee`, `IndexMemberByOrganization`, and `UpdateMember` (with call recording + error simulation). Comment text only — no behavior change. Fixed in `fdd39a4`.

**Skipped:** none.

**Verification (all green):** `gofmt -l` clean · `go vet ./...` · `go build ./...` · `go test -count=1 ./...` (all pass) · `golangci-lint run ./...` → 0 issues.

File: internal/infrastructure/nats/org_committee_seat_reader.go [hunk 2]
```diff
@@ -56,11 +59,25 @@ func (r *natsOrgCommitteeSeatReader) ListOrgCommitteeSeats(ctx context.Context,
...
+	committeeProjectCache := make(map[string]string)
+	recovered := 0
 	filtered := make([]*model.CommitteeMember, 0, len(members))
 	for _, m := range members {
 		if m == nil {
 			continue
 		}
+		if m.ProjectUID == "" {
+			if projectUID := r.resolveCommitteeProjectUID(ctx, m.CommitteeUID, committeeProjectCache); projectUID != "" {
+				m.ProjectUID = projectUID
```
Comment (dealako) [inline internal/infrastructure/nats/org_committee_seat_reader.go]: [minor] In-place mutation of a reader-returned pointer.

`m.ProjectUID = projectUID` mutates the struct that came back from `ListMembersByOrganization`. The NATS storage layer allocates a fresh `*model.CommitteeMember` per KV key on every call, so this is **safe today** — no shared cache to corrupt and downstream consumers receive the enriched value correctly.

That said, this violates the defensive ownership convention: callers shouldn't mutate structs they don't own. If `ListMembersByOrganization` ever adds a per-request or per-org cache, this mutation would silently corrupt it.

Consider a shallow copy before the write:
```go
if m.ProjectUID == "" {
    if projectUID := r.resolveCommitteeProjectUID(ctx, m.CommitteeUID, committeeProjectCache); projectUID != "" {
        enriched := *m   // shallow copy — CommitteeMember has no pointer fields needing deep copy
        enriched.ProjectUID = projectUID
        m = &enriched
        recovered++
    }
}
```
One struct copy per recovered seat (negligible) and makes ownership explicit.

File: cmd/committee-cli/commands/sync/member_project_attribute.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+			"now_project_uid", want.projectUID,
+			"dry_run", rc.DryRun,
+		)
+
+		if rc.DryRun {
+			stats.Updated++
+			return nil
+		}
+
+		revision, errRev := rc.CommitteeReader.GetMemberRevision(ctx, member.UID)
```
Comment (dealako) [inline cmd/committee-cli/commands/sync/member_project_attribute.go]: [minor] TOCTOU between `EachMember` snapshot and `GetMemberRevision`.

The member fields are captured when `EachMember` streams the KV entry. The revision is fetched separately here. If another writer updates the member in that window, `GetMemberRevision` returns the newer revision — the CAS at line 137 (`UpdateMember`) succeeds, but the snapshot fields silently overwrite the concurrent write.

Impact is low (operator-run tool, dry-run default, narrow race window), but worth a comment to warn operators:
```go
// NOTE: revision is fetched after the EachMember snapshot. A concurrent write in this
// window means the CAS succeeds against the newer revision but the snapshot fields
// overwrite it. Run during low-traffic periods to avoid lost updates.
revision, errRev := rc.CommitteeReader.GetMemberRevision(ctx, member.UID)
```

File: cmd/committee-cli/commands/sync/member_project_attribute.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+func (s *memberProjectAttributeSubcommand) Run(ctx context.Context, rc commands.RunContext) error {
+	slog.DebugContext(ctx, "starting subcommand", "subcommand", s.Name(), "args", rc.Args)
+
+	fs := flag.NewFlagSet("member-project-attribute", flag.ContinueOnError)
+	fs.Usage = func() {
+		_, _ = fmt.Fprintf(fs.Output(), "usage: committee-cli sync member-project-attribute [flags]\n\nflags:\n")
+		fs.PrintDefaults()
+	}
+	committeeUID := fs.String("committee-uid", "", "limit repair to members of a single committee UID")
+	projectUID := fs.String("project-uid", "", "limit repair to members whose committee belongs to this project UID (foundation root or child)")
```
Comment (dealako) [inline cmd/committee-cli/commands/sync/member_project_attribute.go]: [nit] Misleading flag description.

`"(foundation root or child)"` implies the flag accepts a foundation root and the repair walks descendants. The check at line 103 is an exact single-project match:
```go
if *projectUID != "" && want.projectUID != *projectUID {
```
Suggested wording:
```go
projectUID := fs.String("project-uid", "", "limit repair to members whose committee belongs to this exact project UID")
```

**Review by dealako — DISMISSED** [review]
@luismoriguerra — nice fix. The root-cause diagnosis is sharp (legacy pre-LFXV2-1442 records silently dropped by the project-family filter), the runtime fallback degrades gracefully, the backfill CLI has safe defaults and idempotency, and the regression test locks the `CreateMember` invariant that was previously untested. The PR is well-scoped and easy to follow.

## AI Comment Reconciliation

- **CodeRabbit (no actionable comments, latest round):** Agree — nothing to add from that review pass.
- **CodeRabbit pre-merge: Docstring coverage 35.29% (threshold 80%):** Disagree that this should block. Exported functions all have proper Go doc comments; the 80% threshold appears misconfigured for a Go codebase. Not a code issue.
- **Copilot (stale `mockMemberWriter` doc comment):** Confirmed resolved in `fdd39a4`. ✅

## Prior Round Resolution

- Round 1 — Copilot raised stale comment on `mockMemberWriter` in `members_by_committee_index_test.go:23` → fixed in `fdd39a4` (comment now accurately describes the full `port.CommitteeMemberWriter` shared across sync backfill tests). ✅

## Issue Count

🔴 **Blocking: 0**
🟡 **Minor: 2** — in-place member mutation (defensive ownership), TOCTOU between EachMember snapshot and GetMemberRevision
⚪ **Nit: 2** — misleading `--project-uid` flag description, missing test for committee-with-empty-project_uid path

## Final Decision

✅ **Approved with minor comments**

**Comment by dealako** [comment]
[nit] Missing test: committee with empty `project_uid`

Both implementations gracefully handle the case where a member's `project_uid` is empty **and** its parent committee's `project_uid` is also empty — the CLI backfill skips with a warning (`member_project_attribute.go:96–101`) and the reader excludes the seat (the empty string never appears in the `family` map). Neither path has a dedicated test.

Suggested addition to `member_project_attribute_test.go`:
```go
func TestMemberProjectAttribute_CommitteeWithNoProjectUID(t *testing.T) {
    r := &mockReader{
        bases: map[string]*model.CommitteeBase{
            "c1": committeeBase("c1", "", ""), // committee itself has no project_uid
        },
        members: map[string][]*model.CommitteeMember{
            "c1": {projMember("m1", "c1", "", "")},
        },
    }
    w := &mockMemberWriter{}
    err := (&memberProjectAttributeSubcommand{}).Run(context.Background(), newBackfillRC(r, w, "--dry-run=false"))
    require.NoError(t, err)
    assert.Empty(t, w.updated, "members whose committee has no project_uid must be skipped, not written")
}
```
Similarly a reader-side test with `byUID["c1"] = ""` would document that the seat is excluded (family never contains `""`) rather than accidentally included.

File: internal/infrastructure/nats/org_committee_seat_reader.go (hunk 2 above)
Comment (luismoriguerra) [reply internal/infrastructure/nats/org_committee_seat_reader.go]: Done in e47f896 — we now enrich a shallow copy (`enriched := *m`) instead of mutating the reader-owned struct, with a comment noting the defensive-ownership rationale (safe against a future per-request/per-org cache). Thanks!

File: cmd/committee-cli/commands/sync/member_project_attribute.go (hunk 3 above)
Comment (luismoriguerra) [reply cmd/committee-cli/commands/sync/member_project_attribute.go]: Added a NOTE in e47f896 documenting the TOCTOU between the `EachMember` snapshot and `GetMemberRevision` (concurrent-write lost-update window; run during low-traffic periods). Thanks!

File: cmd/committee-cli/commands/sync/member_project_attribute.go (hunk 4 above)
Comment (luismoriguerra) [reply cmd/committee-cli/commands/sync/member_project_attribute.go]: Reworded in e47f896 — `--project-uid` help now reads "this exact project UID" to match the single-project match (no foundation-family walk). Thanks!

**Comment by luismoriguerra** [comment]
## Review round 2 — @dealako threads addressed (commit `e47f896`)

| Thread | File | Fix |
|---|---|---|
| In-place mutation of reader-returned pointer | `org_committee_seat_reader.go` | Enrich a **shallow copy** (`enriched := *m`) instead of mutating the struct from `ListMembersByOrganization`; added a comment on the defensive-ownership rationale (safe against a future per-request/per-org cache). |
| TOCTOU: snapshot vs. `GetMemberRevision` | `member_project_attribute.go` | Added a `NOTE` documenting the concurrent-write lost-update window and advising the repair be run during low-traffic periods. |
| Misleading `--project-uid` help | `member_project_attribute.go` | Reworded to "this **exact** project UID" (the filter is a single-project match, not a foundation-family walk). |

All three were comment/doc + a small defensive-copy change — no behavior change to the repair or seat-read logic.

**Verification:** `gofmt -l` clean · `go vet ./...` · `go build ./...` · `go test ./...` pass · `golangci-lint` (v2.2.2) **0 issues**.

All three threads replied-to and marked resolved. (Round 1: the Copilot `mockMemberWriter` doc thread was fixed in `fdd39a4` and resolved.)

File: internal/infrastructure/nats/org_committee_seat_reader_test.go [hunk 5]
```diff
@@ -78,3 +103,51 @@ func TestNATSOrgCommitteeSeatReader_ProjectFamilyFilter(t *testing.T) {
...
+
+		got, err := reader.ListOrgCommitteeSeats(context.Background(), org, []string{"p-1"})
+		require.NoError(t, err)
+		require.Len(t, got, 3)
+
+		uids := map[string]string{}
+		for _, m := range got {
+			uids[m.UID] = m.ProjectUID
+		}
+		assert.Equal(t, "p-1", uids["m-empty-1"], "recovered project_uid must be enriched in place")
```
Comment (dealako) [inline internal/infrastructure/nats/org_committee_seat_reader_test.go]: [nit] Stale assertion message after the shallow-copy refactor.

The message `"recovered project_uid must be enriched in place"` was written when the enrichment mutated `m` directly. Now that `e47f896` uses a shallow copy (`enriched := *m`), the phrase "in place" contradicts the actual behavior — a future reader might interpret it as a claim about in-place mutation and be confused when they see the copy.

Suggested one-word fix:
```go
assert.Equal(t, "p-1", uids["m-empty-1"], "recovered project_uid must be set on the enriched copy")
```

**Review by dealako — DISMISSED** [review]
@luismoriguerra — thanks for the quick and thorough follow-up. Every item from Round 1 is cleanly addressed.

## Revision Tracking

### ✅ Resolved (3/3 from Round 1)

| # | Finding | Resolution |
|---|---------|------------|
| 1 | **[minor]** In-place mutation of reader-returned struct | Fixed in `e47f896` — now enriches a shallow copy (`enriched := *m`; `enriched.ProjectUID = projectUID`; `m = &enriched`) with an explanatory comment about defensive ownership. ✅ |
| 2 | **[minor]** TOCTOU between `EachMember` snapshot and `GetMemberRevision` | Fixed in `e47f896` — 5-line `// TOCTOU:` comment added documenting the lost-update window and recommending low-traffic runs. ✅ |
| 3 | **[nit]** Misleading `--project-uid` flag description | Fixed in `e47f896` — now reads `"this exact project UID"`, accurately reflecting the single-project match. ✅ |

### ⚪ Carried open (1 nit, no elevation)

- **Missing test: committee with empty `project_uid`** (raised as a PR comment, nit-only) — not addressed, which is fine given the severity. The behavior (skip with a warning) is correct and covered by the error-path test. No elevation.

## Security Re-Check

The shallow copy `enriched := *m` was independently verified by the security subagent: `CommitteeMember` has only one pointer field (`Invite *InviteInfo`) which is not mutated by this code path. The copy is safe. TOCTOU documentation is adequate for an operator-run, dry-run-by-default utility. **No security issues.**

## New Issue Scan

One nit on the test assertion message (see inline comment). No new blocking or minor issues.

## AI Comment Reconciliation

- **No new bot comments since Round 1.** The three author replies confirming the fixes are all verified as complete.

## Issue Count (all rounds)

🔴 **Blocking: 0**
🟡 **Minor: 0** — both Round 1 minors resolved
⚪ **Nit: 1** — stale test assertion message ("in place" → post-shallow-copy)

## Final Decision

✅ **Approved**

File: internal/infrastructure/nats/org_committee_seat_reader.go [hunk 6]
```diff
@@ -56,11 +59,31 @@ func (r *natsOrgCommitteeSeatReader) ListOrgCommitteeSeats(ctx context.Context,
...
 	for _, m := range members {
 		if m == nil {
 			continue
 		}
+		if m.ProjectUID == "" {
+			if projectUID := r.resolveCommitteeProjectUID(ctx, m.CommitteeUID, committeeProjectCache); projectUID != "" {
+				// Enrich a shallow copy rather than the reader-owned struct: CommitteeMember has no
+				// pointer fields needing a deep copy, and we never want to mutate what
+				// ListMembersByOrganization handed us (defensive against a future per-request/per-org
+				// cache silently inheriting the enriched value).
```
[the remark this answers is not in the document]
Comment (luismoriguerra) [reply internal/infrastructure/nats/org_committee_seat_reader.go]: You're right — corrected in 1490ec7. `CommitteeMemberBase` does have a pointer field (`Invite *InviteInfo`); the comment now states the shallow copy is safe specifically because we only overwrite the scalar `ProjectUID` and never mutate the shared pointer (so no deep copy is needed). Thanks!

File: cmd/committee-cli/commands/sync/member_project_attribute.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		if rc.DryRun {
+			stats.Updated++
+			return nil
+		}
+
+		// TOCTOU: the member fields were captured by the EachMember snapshot; the revision is read here,
+		// separately. A concurrent write in this window means the CAS below succeeds against the newer
+		// revision but the snapshot fields overwrite it (lost update). The window is narrow and this is an
+		// operator-run, dry-run-by-default repair — run it during low-traffic periods to avoid clobbering
+		// concurrent writes.
```
[the remark this answers is not in the document]
Comment (luismoriguerra) [reply cmd/committee-cli/commands/sync/member_project_attribute.go]: Implemented in 1490ec7. Replaced the EachMember-snapshot + separate `GetMemberRevision` with a single `GetMember` read (member + revision), and now apply ONLY the project fields to that fresh struct — so concurrent edits to other fields are preserved. The revision-checked CAS still guards the read→write window, and I added a fresh-record drift re-check (skips if a concurrent run already repaired it). Test mock `GetMember` updated to resolve by UID. Thanks!

**Comment by luismoriguerra** [comment]
## Review round 3 — @Copilot + @dealako follow-ups addressed (commits `1490ec7`, `a1b72fe`)

| Thread | File | Fix |
|---|---|---|
| Misleading shallow-copy comment | `org_committee_seat_reader.go` | `CommitteeMemberBase` does have a pointer field (`Invite`). Comment corrected: the shallow copy is safe because we only overwrite the scalar `ProjectUID` and never mutate the shared pointer — no deep copy needed. |
| TOCTOU could clobber concurrent writes | `member_project_attribute.go` | Replaced the `EachMember` snapshot + separate `GetMemberRevision` with a single `GetMember` read (member + revision) and apply **only** the project fields to that fresh struct, preserving concurrent edits to other fields. Revision-checked CAS still guards the read→write window, plus a fresh-record drift re-check. Test mock `GetMember` now resolves by UID. |
| Stale "in place" test wording | `org_committee_seat_reader_test.go` | Assertion message + test docstring updated to "set on the enriched copy" / "enriched on a shallow copy" to match the refactor. |

No behavior change to the seat-read result; the backfill now does a proper read-modify-write instead of writing back a stale snapshot.

**Verification:** `gofmt -l` clean · `go vet ./...` · `go build ./...` · `go test ./...` pass · `golangci-lint` (v2.2.2) **0 issues**.

All threads replied-to and resolved. (Rounds 1–2: Copilot `mockMemberWriter` doc + @dealako's shallow-copy / TOCTOU-note / flag-wording threads were resolved in `fdd39a4` and `e47f896`.)

File: internal/infrastructure/nats/org_committee_seat_reader_test.go (hunk 5 above)
Comment (luismoriguerra) [reply internal/infrastructure/nats/org_committee_seat_reader_test.go]: Fixed in a1b72fe — the assertion message now reads "must be set on the enriched copy", and I also updated the test docstring ("enriched on a shallow copy") to match the refactor. Thanks!

**Review by dealako — APPROVED** [review]
@luismoriguerra — 👏 Everything addressed, and then some. The TOCTOU fix went further than I asked: rather than just adding a warning comment, `1490ec7` actually eliminates the lost-update hazard by replacing the stale EachMember snapshot + separate `GetMemberRevision` with a single `GetMember` call that returns both the fresh member state and its revision atomically. Only the project fields are then applied on top of that fresh record, so concurrent edits to other fields are never overwritten. The revision-checked CAS in `UpdateMember` still guards the narrow read→write window.

## Revision Tracking — Round 3

### ✅ All open items resolved

| Round | Finding | Resolution |
|-------|---------|------------|
| R1 [minor] | In-place mutation → defensive ownership | `e47f896`: shallow copy (`enriched := *m`). Comment updated in `1490ec7` to note `Invite *InviteInfo` is a pointer field but safe (never mutated here). ✅ |
| R1 [minor] | TOCTOU: snapshot + separate `GetMemberRevision` | `1490ec7`: **fully eliminated** — replaced with `GetMember` (fresh member + revision together); fresh-record drift re-check added for concurrent-repair idempotency. ✅ |
| R1 [nit] | Misleading `--project-uid` flag description | `e47f896`: "this exact project UID". ✅ |
| R2 [nit] | Stale test assertion `"enriched in place"` | `a1b72fe`: updated to `"must be set on the enriched copy"`; docstring also corrected to `"enriched on a shallow copy"`. ✅ |

### Carried open (still a nit, no elevation)

- **Missing test: committee with empty `project_uid`** — nit from Round 1, not required for merge.

## New Commits — Issue Scan

**`1490ec7`** — no issues.
- `GetMember` mock in `total_members_attribute_test.go` correctly updated to resolve by UID (was stub returning `nil`), so all existing backfill tests now exercise the new code path.
- The fresh-record re-check (`if fresh.ProjectUID == want.projectUID`) is a clean idempotency guard.
- Tests: all pass (`go test ./...`).

**`a1b72fe`** — test wording only, no logic changes.

## Issue Count (all rounds)

🔴 **Blocking: 0** 
🟡 **Minor: 0** 
⚪ **Nit: 0** (all addressed)

## Final Decision

✅ **Approved**
